### PR TITLE
Dev v1.4

### LINKED ISSUE #323: Evaluate addition of nR ratio
First step is probably to break the items from Jim's slides in to multiple issues: 

[nR modified Hy's Law definition.pptx](https://github.com/SafetyGraphics/hep-explorer/files/4658573/nR.modified.Hy.s.Law.definition.pptx)

### PR DESCRIPTION
**Pull request by jwildfire**
# Overview

This PR adds an nR ratio calculation to the hep-explorer and allows users to show r ratio and nR ratio on the x and y axis. Fixes #323.

# Test Notes

We need to QC the nR ratio calculations before release via double programming using the data sets described below. General idea is to have a tester independently re-calculate participant-level nR ratio data and confirm that it matches the derived data in the tool. Would be good to do this in an R script, and perhaps save it in the `/test/ folder in this repo - like @samussiah did for pAlt [here](https://github.com/SafetyGraphics/hep-explorer/tree/master/test). See  [here](https://github.com/SafetyGraphics/hep-explorer/files/4658573/nR.modified.Hy.s.Law.definition.pptx) for the nR ratio formula. 

Three data sets can be downloaded for any version of the chart by clicking the  "↓ Raw Data"  button at the top of the page. 

- Initial data - This is the raw data that was loaded in to the tool. It is one record per person per measure per visit. To download, click the "↓ Raw Data"  button with no keys pressed.
- Derived data - This is the derived data that the tool uses to create the chart. It is one record per person. To download, click the "↓ Raw Data"  button *while holding the shift key*.
- Visit-level nR Ratio data - This is the visit-level nR ratio data as derived by the tool. It is one record per person per visit. The "value" colum has the nR ratio calculation at that time point. To download, click the "↓ Raw Data"  button *while holding the alt key*.

### CODE REVIEW

**Comment by jwildfire** [comment]
@samussiah Thanks for working on this a little while back. Would live to get this merged before the safetyGraphics v2 release. Do you remember where you left this? Any ideas on next steps?

**Comment by samussiah** [comment]
> @samussiah Thanks for working on this a little while back. Would like to get this merged before the safetyGraphics v2 release. Do you remember where you left this? Any ideas on next steps?

@jwildfire I basically reproduced every subject nR Ratio plot in R and compared them with the app...  Let me double check the comparisons and get back to you.  Do you have any suggestions on a more automated approach?  Perhaps writing a little .js function that exposes the nR ratio calculations for comparison in R?

**Comment by jwildfire** [comment]
> Let me double check the comparisons and get back to you. Do you have any suggestions on a more automated approach? Perhaps writing a little .js function that exposes the nR ratio calculations for comparison in R?

I basically did this - from the test notes in the PR above:

> Visit-level nR Ratio data - This is the visit-level nR ratio data as derived by the tool. It is one record per person per visit. The "value" colum has the nR ratio calculation at that time point. To download, click the "↓ Raw Data" button while holding the alt key.

So, seems like we could just pull that .csv, save it to your new validation folder and then do a quick compare.

**Comment by samussiah** [comment]
I'll knock it out shortly.

File: src/callbacks/onInit/cleanData/makeRRatio.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            obj.relative_uln = rratio;
+            obj.relative_baseline = rratio_relative_baseline;
+            // note that the baseline calculation uses raw values and doesn't account for changes in ULN over the course of the study.
+            return obj;
+        })
+        .entries(chart.imputed_data)
+        .map(m => m.values)
+        .filter(f => f.absolute || f.relative_baseline);
+
+    let nrRatio = d3
```
Comment (samussiah) [inline src/callbacks/onInit/cleanData/makeRRatio.js:48]: Perfect.

File: src/callbacks/onLayout/addDownloadButton.js [hunk 2]
```diff
@@ -21,17 +21,36 @@ export function addDownloadButton() {
...
+                            const obj = m.nrRatio_raw;
+                            obj.forEach(function(mm) {
+                                mm.id = m[config.id_col];
+                            });
+                            return obj;
+                        })
+                    );
+                    cols = Object.keys(nrRatioData[0]);
+                    downloadCSV.call(this, nrRatioData, cols, 'eDish_nrRatioData_testing');
+                } else if (d3.event.shiftKey) {
```
Comment (samussiah) [inline src/callbacks/onLayout/addDownloadButton.js:38]: Clever.

**Review by samussiah — APPROVED** [review]
Math looks good, nailed it!

**Pull request merged**